Pre/Post Schedule Tasks System + Cache Architecture Improvments  +  Rs2* utility class extensions\improvments

Voxsylvae opened the pull request:
## **Overview**

This pull request introduces a comprehensive enhancement of the Microbot Plugin Scheduler through a 5-commit series that transforms plugin automation from manual resource management to declarative requirement specification. 

### **Iimplementation History**

1. **f60a4c1e01** - Foundation: Comprehensive pre/post schedule tasks and requirements system
2. **e76054ac0b** - Enhancement: Major scheduler and Grand Exchange system improvements
3. **f89eef4e59** - Robustness: Cancellation support and comprehensive UI components 
4. **0151c23793** - Intelligence: OR requirement modes with enhanced planning algorithms 
5. **9852349dab** - Optimization: Cache architecture overhaul and system consolidation (final commit)

## **Core Infrastructure Changes**

### **Util Package Enhancements**

#### **Cache Architecture Overhaul**
The cache system received a complete redesign focused on thread safety and profile consistency:

- **Rs2CacheManager**: Complete rewrite with `AtomicReference<String>` patterns for thread-safe profile management
- **Rs2Bank**: Enhanced cache validation with automatic profile consistency checks and retry mechanisms
- **Rs2BankData**: Improved cache loading with configurable retry patterns and error recovery
- **Rs2ItemModel**: Enhanced noted/unnoted item ID resolution with better caching

#### **New Utility Models**
Several new utility classes were added to support the requirements system:

- **Rs2FuzzyItem**: Flexible inventory operations supporting charged items and item variations
- **Rs2DropSource**: Combat drop modeling for loot-based requirements
- **Rs2SpawnLocation**: Ground item tracking and spawn prediction
- **Rs2ShopItem/Rs2ShopSource/Rs2ShopType**: Unified shop operation system
- **Rs2WorldUtil**: Enhanced world management with region preferences and selection modes

### **Task and Requirement System Architecture**

The scheduler system introduces a comprehensive task and requirement framework that changes how plugins handle preparation and cleanup operations:

#### **Abstract Task Management System**
- **`AbstractPrePostScheduleTasks`**: Base class providing standardized task execution lifecycle with CompletableFuture-based cancellation support
- **Thread-Safe Execution**: Dedicated executor services for pre-schedule and post-schedule operations with proper resource isolation
- **State Tracking**: Comprehensive monitoring through `TaskExecutionState` with real-time progress feedback
- **Cancellation Infrastructure**: Emergency hotkey support (Ctrl+C) with graceful shutdown patterns across all requirement types

##### **AbstractPrePostScheduleTasks Deep Analysis** (`runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/AbstractPrePostScheduleTasks.java`)

The base class implements a sophisticated task execution framework with comprehensive lifecycle management:

```java
@Slf4j
public abstract class AbstractPrePostScheduleTasks implements AutoCloseable, KeyListener {
    protected final SchedulablePlugin plugin;
    private ScheduledExecutorService postExecutorService;
    private ScheduledExecutorService preExecutorService;
    private CompletableFuture<Boolean> preScheduledFuture;
    private CompletableFuture<Boolean> postScheduledFuture;
    
    // Emergency cancel hotkey support
    private final KeyManager keyManager;
    
    // Centralized state tracking
    @Getter
    private final TaskExecutionState executionState = new TaskExecutionState();
    
    private final LockCondition prePostScheduleTaskLock = new LockCondition("Pre/Post Schedule Task Lock");
}
```

**Key Infrastructure Features:**
- **Dual Executor Architecture**: Separate `ScheduledExecutorService` instances for pre and post tasks prevent execution interference
- **CompletableFuture Lifecycle**: `preScheduledFuture` and `postScheduledFuture` provide timeout support and cancellation capabilities
- **Emergency Cancellation**: Ctrl+C hotkey registration through `KeyManager` enables manual task abortion
- **LockCondition Integration**: Thread-safe locking prevents concurrent task execution conflicts
- **State Validation**: `canStartPreScheduleTasks()` and `canStartPostScheduleTasks()` ensure execution prerequisites

**Thread Safety Patterns:**
```java
private boolean canStartAnyTask(){
    if (plugin == null) {
        log.warn("Plugin instance is null, cannot determine schedule mode");
        return false;
    }
    
    if (getRequirements() == null || !getRequirements().isInitialized()) {
        log.warn("Requirements are not initialized, cannot execute pre-schedule tasks");
        return false;
    }
    
    if (postScheduledFuture != null && !postScheduledFuture.isDone()) {
        log.warn("Post-schedule task is still running, cannot execute pre-schedule tasks yet");
        return false;
    }
    
    return true;
}
```

#### **Requirements Registry Architecture**
- **`RequirementRegistry`**: Centralized requirement management with type-safe access and uniqueness enforcement
- **Context-Aware Execution**: Requirements categorized by `TaskContext` (PRE_SCHEDULE, POST_SCHEDULE, IMMEDIATE)
- **Priority Framework**: `RequirementPriority` system (MANDATORY, RECOMMENDED, OPTIONAL) for intelligent fulfillment ordering
- **Dependency Resolution**: Automatic analysis of requirement relationships for optimal execution sequences

#### **Execution Result Framework**
Replace boolean success patterns with granular `ExecutionResult` enum system:
- **SUCCESS**: Plugin completed successfully without issues, ready for normal scheduling
- **SOFT_FAILURE**: Recoverable issues allowing retry (temporary resource unavailability, network timeouts)
- **HARD_FAILURE**: Critical failures preventing future execution (invalid configuration, missing dependencies)

## **Plugin Scheduler Core Changes**

### **SchedulablePlugin API Evolution**

The SchedulablePlugin interface received major enhancements for better lifecycle management:

#### **New Event System**
- **ExecutionResult Enum**: Replaces boolean success patterns with granular results (SUCCESS, SOFT_FAILURE, HARD_FAILURE)
- **Enhanced Event Handlers**: New pre/post schedule task events with detailed execution feedback
- **Default Implementations**: Comprehensive default methods for common plugin lifecycle patterns
- **Improved Error Reporting**: Standardized error handling with actionable failure information

#### **Enhanced Lifecycle Management**
```java
// New execution result patterns
public ExecutionResult executePreScheduleTasks();
public ExecutionResult executePostScheduleTasks();

// Enhanced event handling
@Subscribe
public void onPluginScheduleEntryPreScheduleTaskEvent(PluginScheduleEntryPreScheduleTaskEvent event);
@Subscribe  
public void onPluginScheduleEntryPostScheduleTaskEvent(PluginScheduleEntryPostScheduleTaskEvent event);
```

#### **Configuration Integration**
- **ConfigDescriptor Support**: Enhanced configuration management for per-entry settings
- **Pre/Post Tasks Integration**: Seamless integration with AbstractPrePostScheduleTasks system
- **Custom Requirement Support**: External requirement registration capabilities

### **SchedulerPlugin Core Updates**

- **Plugin Management**: Automatic disabling of non-essential plugins during scheduler startup for better resource management
- **Enhanced State Tracking**: Improved PluginScheduleEntry with better execution tracking and failure handling
- **Plugin Filtering**: New PluginFilterUtil for intelligent plugin categorization and management
- **Lock State Management**: Enhanced lock conditions for preventing interruptions during critical operations

## **Plugin Integration Implementation**

### **GOTR (Guardians of the Rift) Integration**

The GOTR plugin demonstrates comprehensive task and requirement system integration for complex minigame automation:

#### **Requirements Implementation** (`GotrPrePostScheduleRequirements.java`)
- **Minigame-Specific Requirements**: Optimal equipment setups including Runecrafting outfit (Robes of the Eye), Graceful outfit, and mining equipment
- **World Selection Logic**: Automatic selection of most populated GOTR world from dedicated world list (409, 445, 464, 478, 522, 534)
- **Conditional Spellbook Management**: Lunar spellbook requirement for NPC Contact spell with automatic restoration
- **Location Requirements**: Pre-schedule positioning at GOTR bank, post-schedule movement to Grand Exchange
- **ItemRequirementCollection Integration**: Standardized outfit registration using collection utilities

#### **Task Management** (`GotrPrePostScheduleTasks.java`)
- **Skill Level Validation**: Automatic verification of Runecrafting level requirements before execution
- **Quest Prerequisite Checking**: Validation of Temple of the Eye quest completion
- **Custom Preparation Logic**: GOTR-specific setup including pouch preparation and inventory optimization
- **Resource Cleanup**: Comprehensive cleanup handling for minigame-specific items and states

### **Agility Plugin Integration**

The Agility plugin showcases sophisticated requirement planning with conditional logic and multi-item configurations:

#### **Requirements Implementation** (`MicroAgilityPrePostScheduleRequirements.java`)
- **Conditional Alchemy Requirements**: Dynamic requirement generation based on available items and alchemy configuration
- **Fire Rune vs Fire Staff Logic**: Intelligent conditional requirements that automatically select between fire runes and fire staves based on availability
- **Item Availability Detection**: Real-time scanning of inventory and bank for alchable items with quantity-based requirement generation
- **Graceful Outfit Integration**: Standardized outfit requirements using ItemRequirementCollection
- **Course Location Requirements**: Dynamic location requirements based on selected agility course configuration

#### **Advanced Conditional Logic**
```java
ConditionalRequirement fireAlchRequirement = new ConditionalRequirement()
    .addStep(
        () -> hasFireStaffAvailable(fireStaffRequirements),
        fireStaffOrRequirement,
        "Fire staff available - equip fire staff"
    )
    .addStep(
        () -> !hasFireStaffAvailable(fireStaffRequirements),
        fireRuneRequirement,
        "No fire staff available - get fire runes"
    );
```

#### **Task Management** (`MicroAgilityPrePostScheduleTasks.java`)
- **Configuration Validation**: Comprehensive validation of alchemy settings and item configurations
- **Dynamic Requirement Adjustment**: Real-time modification of requirements based on current game state
- **Post-Schedule Location Management**: Automatic movement to Grand Exchange for banking and trading

## **Revolutionary Requirements System**

### **Declarative Requirement Framework**
The requirements system introduces a paradigm shift from imperative "how-to" programming to declarative "what-I-need" specification, enabling plugins to specify their needs without implementing fulfillment logic.

#### **Core Requirement Architecture**
All requirements extend the abstract `Requirement` base class which provides:
- **Priority Framework**: `RequirementPriority` system (MANDATORY, RECOMMENDED, OPTIONAL)
- **Task Context**: Schedule-aware execution (`PRE_SCHEDULE`, `POST_SCHEDULE`, `IMMEDIATE`)
- **Rating System**: 1-10 effectiveness scoring for intelligent selection
- **Unique Identification**: Automatic uniqueness enforcement preventing conflicts

```java
public abstract class Requirement implements Comparable<Requirement> {
    protected final RequirementType requirementType;
    protected final RequirementPriority priority;
    protected final String description;
    protected final List<Integer> ids; // Item IDs, NPC IDs, etc.
    protected TaskContext taskContext;
    
    // Core contract methods
    public abstract String getName();
    public abstract boolean fulfillRequirement(CompletableFuture<Boolean> scheduledFuture);
    public abstract boolean isFulfilled();
}
```

#### **Comprehensive Requirement Types**

**ItemRequirement**: Advanced item management with intelligent resolution
- **Fuzzy Matching**: Support for charged items (`ring of dueling(8)` → `ring of dueling(7)`)
- **Auto-Resolution**: Noted items automatically resolved to unnoted equivalents
- **Skill Validation**: Level requirements for equipment and usage
- **Equipment Slot Management**: Automatic assignment to optimal slots
- **Type Inference**: INVENTORY, EQUIPMENT, or EITHER based on slot parameters

```java
// Equipment requirement with skill validation
ItemRequirement weapon = ItemRequirement.forEquipment(
    "Dragon scimitar", 
    EquipmentInventorySlot.WEAPON,
    RequirementPriority.MANDATORY,
    TaskContext.PRE_SCHEDULE
).withSkillRequirement(Skill.ATTACK, 60);

// Inventory requirement with fuzzy matching for charged items
ItemRequirement teleports = ItemRequirement.forInventory(
    "Ring of dueling", 8, // Any charge level accepted
    RequirementPriority.RECOMMENDED,
    TaskContext.PRE_SCHEDULE
).withFuzzyMatching(true);
```

**LocationRequirement**: Intelligent location management with pathfinding
- **Multi-Location Support**: Alternative locations with intelligent selection
- **Transport Integration**: Automatic teleport and transport usage
- **Proximity Optimization**: `getBestAvailableLocation()` for efficiency
- **World-Specific Requirements**: Automatic world hopping when needed

```java
LocationRequirement grandExchange = LocationRequirement.atLocation(
    BankLocation.GRAND_EXCHANGE.getWorldPoint(),
    10 // tiles radius
).withWorldHopping(true)
 .withTransports(true);
```

**ShopRequirement**: Advanced trading system with market integration
- **Multi-Item Operations**: Batch processing for efficiency
- **Grand Exchange Integration**: WikiPrice API for real-time pricing
- **World-Hopping Logic**: Automatic world selection for stock availability
- **Offer Recovery**: Intelligent cancellation and recovery of failed offers
- **Progress-Based Prioritization**: Sophisticated offer management

```java
ShopRequirement tradingReq = ShopRequirement.createBuyOrder()
    .addItem("Maple longbow", 100, 5000) // item, quantity, max price
    .addItem("Yew longbow", 50, 8000)
    .withWorldHopping(true)
    .withRetryMechanisms(3);
```

**LootRequirement**: Ground item collection with clustering algorithms
- **Spawn Location Tracking**: Intelligent spawn point monitoring
- **Cluster Analysis**: Group nearby spawn locations for efficiency
- **Timeout Management**: Configurable collection time limits
- **Multi-Item Support**: Collect different items from same area

```java
LootRequirement lootReq = LootRequirement.builder()
    .addItem("Bones", 10)
    .addItem("Raw beef", 5)
    .withSpawnLocation(spawnLocation)
    .withTimeout(Duration.ofMinutes(5))
    .withClusterProximity(15); // tiles
```

#### **Logical Requirement Composition**

**OrRequirement**: Alternative requirement specifications with intelligent algorithms
- **ANY_COMBINATION Mode**: Maximum flexibility for mixed fulfillment
- **SINGLE_TYPE Mode**: Focused fulfillment preferring single items
- **Smart Selection**: Automatic best-option detection based on availability

**ConditionalRequirement**: State-dependent requirement activation
- **Dynamic Conditions**: Requirements that activate based on game state
- **Chain Support**: Multi-step conditional logic with fallbacks
- **Context Awareness**: Integration with current player status

**OrderedRequirement**: Sequential fulfillment when order matters
- **Dependency Management**: Automatic prerequisite handling
- **State Preservation**: Maintains progress through complex sequences
- **Rollback Support**: Recovery from partial failures

#### **RequirementRegistry Architecture**
Centralized requirement management with uniqueness enforcement:
- **Thread-Safe Operations**: ConcurrentHashMap-based storage
- **Automatic Deduplication**: UniqueIdentifier-based key system
- **Efficient Lookups**: Cached views by type and context
- **External Integration**: Separate namespace for plugin-added requirements

```java
public class RequirementRegistry {
    // Central storage with uniqueness enforcement
    private final Map<RequirementKey, Requirement> requirements;
    
    // Cached views for efficient access
    private volatile Map<TaskContext, Map<EquipmentInventorySlot, 
        LinkedHashSet<ItemRequirement>>> equipmentItemsCache;
    
    // Type-safe accessors
    public List<ItemRequirement> getEquipmentRequirements(TaskContext context);
    public List<LocationRequirement> getLocationRequirements(TaskContext context);
    public List<ShopRequirement> getShopRequirements(TaskContext context);
}
```

### **Advanced Planning Algorithms**

#### **OR Requirement Modes**
Revolutionary requirement planning with two distinct strategies:
- **`ANY_COMBINATION`**: Maximum flexibility allowing mixed item fulfillment (e.g., "2 lobsters + 3 swordfish")
- **`SINGLE_TYPE`**: Focused fulfillment preferring single item types (e.g., "5 lobsters OR 5 swordfish")

#### **InventorySetupPlanner Integration**

The `InventorySetupPlanner` represents a major advancement in automatic inventory management, directly generating `InventorySetup` configurations from requirement specifications.

**Core Architecture:**
```java
public class InventorySetupPlanner {
    // Equipment slot assignments (fixed positions)
    private final Map<EquipmentInventorySlot, ItemRequirement> equipmentAssignments;
    
    // Specific inventory slot assignments (0-27)
    private final Map<Integer, ItemRequirement> inventorySlotAssignments;
    
    // Flexible inventory items that can be placed in any available slot
    private final List<ItemRequirement> flexibleInventoryItems;
    
    // Requirements planning support
    private RequirementRegistry registry;
    private TaskContext taskContext;
    private OrRequirementMode orRequirementMode;
}
```

**Requirements-Based Planning:**
- **Automatic Layout Generation**: Converts requirement specifications into optimized inventory layouts
- **Slot Constraint Management**: Respects equipment slot requirements and inventory slot preferences
- **Stackability Intelligence**: Automatic handling of stackable vs non-stackable item placement
- **Missing Item Detection**: Comprehensive analysis of mandatory vs optional missing requirements

**MInventorySetupsPlugin Integration:**
The planner seamlessly integrates with the existing MInventorySetupsPlugin system:
- **Direct InventorySetup Generation**: Creates compatible setup configurations for existing UI
- **Layout Preservation**: Maintains user-defined layouts while optimizing for requirements
- **Variation Mapping Support**: Leverages `InventorySetupsVariationMapping` for item equivalents
- **Export/Import Compatibility**: Generated setups work with all existing import/export features

**Smart Planning Features:**
```java
// Requirements-based constructor for advanced planning
InventorySetupPlanner planner = new InventorySetupPlanner(
    registry,                          // Requirement registry
    TaskContext.PRE_SCHEDULE,         // Schedule context
    OrRequirementMode.ANY_COMBINATION // Planning strategy
);

// Automatic generation from requirements
InventorySetup generatedSetup = planner.generateInventorySetup(
    "Generated Setup for " + pluginName,
    true  // includeOptionalItems
);

// Integration with existing MInventorySetupsPlugin
MInventorySetupsPlugin.importSetup(generatedSetup);
```

**Advanced Conflict Resolution:**
- **Priority-Based Assignment**: Mandatory requirements take precedence over optional ones
- **Slot Competition Handling**: Intelligent resolution when multiple items compete for same slots
- **Overflow Management**: Automatic handling when requirements exceed available inventory space
- **Fallback Strategies**: Alternative placement when preferred slots are unavailable

## **Enhanced Grand Exchange Integration**

### **Multi-Item Trading System**
- **`MultiItemConfig`**: Sophisticated configuration for complex trading scenarios (maple bow → maple longbow → NPC shop)
- **`CancelledOfferState`**: Intelligent offer recovery with selective slot cancellation and restoration
- **World-Hopping Integration**: Automatic world selection for optimal shop availability and trading conditions
- **Progress-Based Cancellation**: Priority-based offer cancellation considering completion progress

### **Real-Time Market Integration**
- **WikiPrice API Integration**: Live market data for adaptive buy/sell pricing strategies
- **`TimeSeriesAnalysis`**: Historical price tracking for intelligent trading decisions
- **Adaptive Pricing Mechanisms**: Dynamic price adjustment based on market conditions and success rates
- **`GrandExchangeOfferDetails`**: Enhanced offer tracking with comprehensive state management

## **Comprehensive UI Framework**

### **Task Monitoring Components**
- **`PrePostScheduleTasksInfoPanel`**: Real-time task execution monitoring with progress indicators
- **`TaskExecutionStatePanel`**: Detailed execution feedback with step-by-step progress tracking
- **`RequirementsStatusPanel`**: Comprehensive requirement fulfillment status with validation summaries
- **`SchedulerInfoPanel` Integration**: Unified task visibility across the entire scheduler ecosystem

### **User Experience**
- **Cancellation Hotkey Support**: Manual task cancellation via configurable hotkeys (`cancelTasksHotkey`)
- **Real-Time Progress Feedback**: Live updates on requirement fulfillment progress and estimated completion times
- **Comprehensive Error Reporting**: Detailed failure analysis with actionable recovery suggestions
- **Overlay Component Integration**: Seamless integration with existing plugin overlay systems

## *Significant API Changes**

### **SchedulablePlugin Interface**
Enhanced plugin integration with sophisticated lifecycle management:

```java
// Enhanced event handling with ExecutionResult
@Subscribe
public void onPluginScheduleEntryPreScheduleTaskEvent(PluginScheduleEntryPreScheduleTaskEvent event);

@Subscribe  
public void onPluginScheduleEntryPostScheduleTaskEvent(PluginScheduleEntryPostScheduleTaskEvent event);

// New execution result patterns
public ExecutionResult executePreScheduleTasks();
public ExecutionResult executePostScheduleTasks();
```

### **Breaking Changes with Migration Paths**
- **Method Signature Updates**: All `fulfillRequirement()` methods now require `CompletableFuture<Boolean>` parameters
- **Priority System Refactoring**: `Priority` enum renamed to `RequirementPriority` for clarity and namespace consistency
- **Package Reorganization**: Requirements moved to logical packages (`item/`, `shop/`, `collection/`, `logical/`)
- **Collection Naming**: `RequirementCollections` renamed to `ItemRequirementCollection` for specificity

## * Utility System Enhancements**

### **Skills Package Reorganization**
Complete restructuring of skill-related utilities for better organization and maintainability:

```
util/skills/
├── fishing/     # Fishing utilities, spot locations, fish types
├── mining/      # Mining rock locations, efficiency calculations  
├── slayer/      # Slayer task management, equipment requirements
└── woodcutting/ # Tree locations, efficiency algorithms
```

### **Resource Location**
- **`ResourceLocationOption`**: Replace `LocationOption` with resource-aware location selection
- **Efficiency Scoring**: Intelligent location ranking based on resource density and player proximity
- **Dynamic Availability**: Real-time resource count monitoring for optimal location selection
- **Proximity Optimization**: Advanced algorithms for minimizing travel time and maximizing efficiency

### **New Utility Models**
- **`Rs2ShopItem`/`Rs2ShopSource`/`Rs2ShopType`**: Unified shop operation system
- **`Rs2DropSource`**: Combat drop modeling for loot-based requirements
- **`Rs2SpawnLocation`**: Ground item tracking and spawn prediction
- **`Rs2FuzzyItem`**: Flexible inventory operations with intelligent item matching

## **Comprehensive Documentation**

### **Some Documentation...**
- **`docs/scheduler/tasks/README.md`**: Overview of the entire task system architecture
- **`docs/scheduler/tasks/task-management-system.md`**: Deep dive into infrastructure and execution patterns
- **`docs/scheduler/tasks/requirements-system.md`**: Comprehensive requirement type documentation
- **`docs/scheduler/tasks/enhanced-schedulable-plugin-api.md`**: Advanced integration patterns and best practices
- **`docs/scheduler/tasks/plugin-writers-guide.md`**: Step-by-step implementation guidance
- **`docs/scheduler/tasks/requirements-integration.md`**: Strategies for requirement system usage beyond scheduler context

### **Example Implementations**
- **GOTR Integration**: implementation demonstrating minigame automation
- **SchedulableExample Plugin**: reference implementation with maple bow trading workflows
- **Fishing/Mining/Woodcutting Requirements**: Skill-specific requirement templates for common activities, future, use it to integrate the task system

## **Alpha/Beta Feature Status**

### **Alpha Features (Experimental - Use with Caution)**


**Multi-Item Shop Requirements**
- **Status**: Alpha with limited Grand Exchange integration
- **Description**: Sophisticated Grand Exchange trading workflows with WikiPrice API integration
- **Recommendation**: Monitor trades manually, test with small quantities first


**LootRequirement Ground Item Collection**
- **Status**: Alpha with basic spawn tracking
- **Description**: Automated ground item collection with cluster analysis and spawn prediction
- **Recommendation**: Use in low-competition areas, supplement with manual validation

### **Beta Features (Stable Testing Phase)**

**World-Hopping Integration**
- **Status**: Alpha implementation for specific shop and activity requirements
- **Description**: Automatic world selection for optimal resource availability and reduced competition
- **Recommendation**: Use conservative hopping intervals, monitor for unusual patterns

**InventorySetupPlanner Automatic Generation**
- **Status**: Beta with comprehensive testing across multiple plugins
- **Description**: Automatic inventory setup generation directly from requirement specifications
- **Validation**: Tested with GOTR, Agility, and fishing plugin requirements
- **Performance**: Reliable for standard use cases, handles most stackability edge cases
- **Integration**: Fully compatible with existing MInventorySetupsPlugin infrastructure

**OR Requirement Planning Algorithms**
- **Status**: Beta with extensive validation
- **Description**: Advanced algorithm selection between ANY_COMBINATION and SINGLE_TYPE fulfillment modes
- **Validation**: Comprehensive testing with food combinations, equipment alternatives, resource gathering
- **Performance**: Optimized selection algorithms, handles complex requirement trees efficiently
- **Edge Cases**: Well-tested with conflicting priorities and resource constraints

**Enhanced Grand Exchange Integration**
- **Status**: Beta with market data integration
- **Description**: Real-time market data integration with historical price analysis
- **Validation**: Tested with maple bow trading workflows and various commodity trading
- **Performance**: Reliable price detection, adaptive buy/sell strategies
- **API Integration**: Stable WikiPrice API connection with fallback mechanisms

**Comprehensive Task Execution Framework**
- **Status**: Beta with production stability
- **Description**: AbstractPrePostScheduleTasks infrastructure with CompletableFuture lifecycle management
- **Validation**: Extensively tested across all enhanced plugins
- **Performance**: Thread-safe execution, proper resource cleanup, emergency cancellation support
- **Reliability**: Comprehensive error handling, timeout management, state recovery


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Pre/Post schedule tasks system: declarative requirements, orchestrated pre/post tasks, overlays, hotkeys and scheduler-driven plugin lifecycle.

* **Changes**
  * Scheduler UI improvements: sanitized plugin names, conditional break info and richer task/status overlays.
  * Advanced world-selection and login watchdog with extended-sleep handling.
  * Public scheduler event/API renames (soft-stop → post-schedule-task, finished → main-task-finished) and richer ExecutionResult statuses.

* **Bug Fixes**
  * Safer UI/EDT handling, cache-overlay gating, stronger null/precondition checks, improved locking/unlock-on-shutdown and alive reporting.

* **Documentation**
  * Expanded guides for tasks, requirements, plugin-writer integration and API migration notes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Voxsylvae committed:
feat: Implement comprehensive pre/post schedule tasks and requirements system

- Add unified requirement framework with conditional logic
- Implement ItemRequirement, LocationRequirement, SpellbookRequirement, etc.
- Create pre/post schedule task execution system with state tracking
- Add InventorySetupPlanner for complex item preparations
- Implement Rs2FuzzyItem for flexible inventory operations
- Add comprehensive examples and plugin integrations
- Create requirement solver with logical operators (AND/OR)
- Add skill-specific requirement templates (Fishing, Mining, Woodcutting)
- Enhance GOTR plugin with schedulable task support
- Add VoxPlugins example implementation with unified location system

Voxsylvae committed:
feat: enhance pre/post schedule tasks with cancellation support and comprehensive UI

Technical enhancements:
- Replace ScheduledExecutorService with CompletableFuture<Boolean> for proper task cancellation
- Add cancellation checks throughout requirement fulfillment chain
- Implement graceful shutdown patterns for all requirement types

Shop system improvements:
- Add CancelledOfferState for Grand Exchange offer recovery
- Implement selective slot cancellation with restoration capability
- Extract MultiItemConfig and ShopOperation models for better modularity
- Enhance GE slot management with progress-based cancellation priority

UI components added:
- PrePostScheduleTasksInfoPanel for comprehensive task monitoring
- TaskExecutionStatePanel for real-time execution feedback
- RequirementsStatusPanel for fulfillment progress tracking
- Integration with SchedulerInfoPanel for unified task visibility

Core system updates:
- Add EXTERNAL_REQUIREMENTS step to FulfillmentStep enum
- Enhance requirement solver with cancellation-aware patterns
- Update all requirement types to support CompletableFuture cancellation
- Improve error handling and resource cleanup across task execution

Configuration enhancements:
- Add cancelTasksHotkey for manual task cancellation
- Streamline overlay components and remove unused imports
- Update example plugin with cancellation demonstration features

Breaking changes:
- fulfillRequirement() method signature updated across all requirement classes
- Task execution patterns now require CompletableFuture for proper lifecycle management

Voxsylvae committed:
feat: Major enhancements to Pre/Post Scheduler and Grand Exchange systems

Core Changes:
- Restructured requirement system with logical package organization
- Implemented comprehensive multi-item shop trading system
- Integrated real-time Wiki API pricing with adaptive strategies
- Added time-series market analysis capabilities

Requirement System Refactoring:
- Reorganized requirements into item/, shop/, collection/, logical/ packages
- Renamed RequirementCollections → ItemRequirementCollection
- Added new ShopRequirement with multi-item support and world-hopping
- Enhanced LocationRequirement with getBestAvailableLocation()

Grand Exchange Enhancements:
- Added WikiPrice API integration for real-time market data
- Implemented TimeSeriesAnalysis for historical price tracking
- Created adaptive buy/sell pricing with retry mechanisms
- Added GrandExchangeOfferDetails for enhanced offer tracking

New Models and Utilities:
- Rs2ShopItem/Rs2ShopSource/Rs2ShopType for unified shop operations
- Rs2DropSource for combat drop modeling
- Rs2SpawnLocation for ground item tracking
- TimeSeriesDataPoint/TimeSeriesInterval for market analysis

Example Implementation:
- SchedulableExample updated with maple bow trading example
- Demonstrates GE buy → NPC shop sell workflow
- Shows multi-item shop requirement configuration

This update provides a robust foundation for advanced trading automation
and intelligent requirement fulfillment with real market data integration.

Voxsylvae committed:
feat: implement OR requirement modes with enhanced planning system

- Add OrRequirementMode enum (ANY_COMBINATION vs SINGLE_TYPE)
- Enhance InventorySetupPlanner with requirements-based constructor
- Add comprehensive OR requirement planning algorithms
- Implement mode-specific planning (mixed items vs single type)
- Refactor PrePostScheduleRequirements to delegate to enhanced planner
- Add detailed requirement analysis and logging system
- Maintain backward compatibility with default ANY_COMBINATION mode

Example: Food requirements can now be fulfilled with '5 lobsters OR 5 swordfish'
(SINGLE_TYPE) instead of allowing '2 lobsters + 3 swordfish' (ANY_COMBINATION)."

refactor: reorganize skill utilities and enhance resource location system

- Move slayer utilities from util/slayer to util/skills/slayer
- Move woodcutting utilities from util/woodcutting to util/skills/woodcutting
- Move mining utilities to util/skills/mining for consistency
- Replace LocationOption with ResourceLocationOption for resource tracking
- Add resource count and efficiency scoring to location selection
- Update all import statements and references across codebase
- Enhance woodcutting location data with resource availability
- Improve location selection algorithms for optimal efficiency
- Add fishing utilities and requirements to new skills structure

This restructuring provides better code organization and more intelligent
resource location selection based on availability and player proximity.

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/VoxPlugins/schedulable/example/SchedulableExamplePlugin.java:
```diff
@@ -132,22 +232,74 @@ protected void startUp() {
+            
+        
             
+            try {
+                // Execute pre-schedule tasks with callback to start the script
+                runPreScheduleTasks();
+            } catch (Exception e) {
+                log.error("Error during Pre-Schedule Tasks for SchedulableExample", e);                
+            }
+        }
```
executePreScheduleTasks in handlePreTaskCompletion of the AbstractPrePostScheduleTasks.java,  handles the report already

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/condition/ui/ConditionConfigPanel.java:
```diff
@@ -368,6 +372,45 @@ private List<Condition> getCurrentConditions() {
+        
+        return true;
+    }
+    
```
can be ignored for now, will be in later pull request

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/examples/ExamplePrePostScheduleRequirements.java:
```diff
@@ -0,0 +1,151 @@
+
+import net.runelite.api.EquipmentInventorySlot;
+import net.runelite.api.Skill;
+import net.runelite.api.gameval.ItemID;
```
we want to use the new item ids , in @net.runelite.api.gameval.ItemID

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/examples/ExamplePrePostScheduleTasks.java:
```diff
@@ -0,0 +1,247 @@
+@Slf4j
+public class ExamplePrePostScheduleTasks extends AbstractPrePostScheduleTasks {
+    
+    private ExamplePrePostScheduleRequirements exampleRequirements;
+
+    /**
+     * Constructor for ExamplePrePostScheduleTasks.
+     * 
+     * @param plugin The SchedulablePlugin instance to manage
+     */
+    public ExamplePrePostScheduleTasks(SchedulablePlugin plugin) {
+        super(plugin,null);
+        this.exampleRequirements = new ExamplePrePostScheduleRequirements();
+    }
```
its just another example, requirement implantation , not related SchedulableExamplePrePostScheduleRequirements

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/AbstractPrePostScheduleTasks.java:
```diff
@@ -0,0 +1,1096 @@
+            
+            // Return true if either succeeded (cleanup is more lenient than setup)
+            return customTasksSuccessful && standardRequirementsFulfilled;
+            
```
Wrong comment, either is both.

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/requirement/conditional/OrderedRequirement.java:
```diff
@@ -0,0 +1,427 @@
+        allStepsCompleted = true;
+        log.debug("Ordered requirement completed successfully: {}", getName());
+        return true;
+    }
```
Good suggestion. add it to the abstract runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/requirement/Requirement.java class, so every requirement can use it

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/data/ItemRequirementCollection.java:
```diff
@@ -0,0 +1,1450 @@
+     */
+    public static void registerRunePouches(PrePostScheduleRequirements requirements, RequirementPriority priority,TaskContext taskContext) {
+        // Rune pouches for efficient rune storage
+       requirements.register(ItemRequirement.createOrRequirement(
+            Arrays.asList(ItemID.RCU_POUCH_COLOSSAL, ItemID.RCU_POUCH_COLOSSAL_DEGRADE),1,
+            -1, priority, 10, "Colossal pouch (for maximum essence carrying)",taskContext
+        ));
+        requirements.register(ItemRequirement.createOrRequirement(
+            Arrays.asList(ItemID.RCU_POUCH_GIANT, ItemID.RCU_POUCH_GIANT_DEGRADE),1,
+            -1, priority, 8, "Giant pouch (for high essence carrying)",taskContext
+        ));
+        requirements.register(ItemRequirement.createOrRequirement(
+            Arrays.asList(ItemID.RCU_POUCH_LARGE, ItemID.RCU_POUCH_LARGE_DEGRADE),1,
+            -1, priority, 6, "Large pouch (for good essence carrying)",taskContext
+        ));
+        requirements.register(ItemRequirement.createOrRequirement(
+            Arrays.asList(ItemID.RCU_POUCH_MEDIUM, ItemID.RCU_POUCH_MEDIUM_DEGRADE),1,
+            -1, priority, 4, "Medium pouch (for decent essence carrying)",taskContext
+        ));
+        requirements.register(new ItemRequirement(
+            ItemID.RCU_POUCH_SMALL,
+            1, -1, priority, 2, "Small pouch (basic essence carrying)",taskContext
+        ));
+    }
```
The -1 means, the pouch can be placed in any inventory slot of the inventory during the planning phase in  via the runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/PrePostScheduleRequirements.java and in runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/requirement/item/ItemRequirement.java

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/MicroAgilityPrePostScheduleRequirements.java:
```diff
@@ -0,0 +1,368 @@
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.EquipmentInventorySlot;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.gameval.ItemID;
```
we want to use the new classed for game ids #net.runelite.api.gameval.ItemID ids -> always use game id class form runelite-api/src/main/java/net/runelite/api/gameval, the old one in runelite-api/src/main/java/net/runelite/api are Deprecated

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/MicroAgilityPrePostScheduleRequirements.java:
```diff
@@ -0,0 +1,368 @@
+                8,
+                "Nature runes for alchemy (" + totalItemsToAlch + " needed)",
+                TaskContext.PRE_SCHEDULE
+            ));
```
we want to use the new classes for game ids #net.runelite.api.gameval.ItemID ids -> always use game id class form runelite-api/src/main/java/net/runelite/api/gameval, the old one in runelite-api/src/main/java/net/runelite/api are Deprecated -> correctly ids are used here see #net.runelite.api.gameval.ItemID
change to directly register natItemRequirement

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/VoxPlugins/schedulable/example/SchedulableExamplePrePostScheduleRequirements.java:
```diff
@@ -0,0 +1,641 @@
+             StoreLocations nearestStore = StoreLocations.getNearestStoreWithAllItems(ItemID.HAMMER, ItemID.BUCKET_EMPTY);
+             log.info("Nearest general store with hammer and bucket: {}", nearestStore != null ? nearestStore.getName() : "None found");
+             if (nearestStore != null) {
+                 // Create shop items for hammer and bucket from the nearest general store
+                 WorldArea storeArea = new WorldArea(
+                     nearestStore.getLocation().getX() - 3, 
+                     nearestStore.getLocation().getY() - 3, 
+                     6, 6, 
+                     nearestStore.getLocation().getPlane()
+                 );
+                 
+                 Rs2ShopItem hammerShopItem = new Rs2ShopItem(
+                     ItemID.HAMMER,
+                     nearestStore.getNpcName(),
+                     storeArea,
+                     nearestStore.getShopType(),
+                     nearestStore.getSellRate(), // Standard sell rate for general stores
+                     nearestStore.getBuyRate(),
+                     nearestStore.getChangePercent(),
+                     nearestStore.getQuestRequirements(),
+                     nearestStore.isMembers(),
+                     "Hammer from " + nearestStore.getName(),
+                     Duration.ofMinutes(5),
+                     5 // Base stock for hammers
+                 );
+                 
+                 Rs2ShopItem bucketShopItem = new Rs2ShopItem(
+                     ItemID.BUCKET_EMPTY,
+                     nearestStore.getNpcName(),
+                     storeArea,
+                     nearestStore.getShopType(),
+                     nearestStore.getSellRate(), // Standard sell rate for general stores
+                     nearestStore.getBuyRate(),
+                     nearestStore.getChangePercent(),
+                     nearestStore.getQuestRequirements(),
+                     nearestStore.isMembers(),
+                     "Empty bucket from " + nearestStore.getName(),
+                     Duration.ofMinutes(5),
+                     3 // Base stock for buckets
+                 );
```
we want to use the new classed for game ids #net.runelite.api.gameval.ItemID ids -> always use game id class form runelite-api/src/main/java/net/runelite/api/gameval, the old one in runelite-api/src/main/java/net/runelite/api are Deprecated

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/MicroAgilityPrePostScheduleRequirements.java:
```diff
@@ -0,0 +1,368 @@
+            8,
+            "Fire runes for alchemy (" + (totalItemsToAlch * 5) + " needed)",
+            TaskContext.PRE_SCHEDULE
+        );
```
we want to use the new classed for game ids #net.runelite.api.gameval.ItemID ids -> always use game id class form runelite-api/src/main/java/net/runelite/api/gameval, the old one in runelite-api/src/main/java/net/runelite/api are Deprecated

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/event/PluginScheduleEntryPostScheduleTaskFinishedEvent.java:
```diff
@@ -0,0 +1,53 @@
+package net.runelite.client.plugins.microbot.pluginscheduler.event;
+
+import java.time.ZonedDateTime;
+
+import lombok.Getter;
+import net.runelite.client.plugins.Plugin;
+
```
ExecutionResult is in the same package

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/SchedulerPlugin.java:
```diff
@@ -2005,11 +2230,139 @@ public Duration getCurrentBreakDuration() {
+                isBreakHandlerEnabled(),
+                Microbot.isLoggedIn(),
+                Microbot.isHopping());
             return;
```
isBreakHandlerEnabled is already checked in isOnBreak

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/PrePostScheduleRequirements.java:
```diff
@@ -0,0 +1,1389 @@
+                            log.info("\\tDeposited individual equipment item: {} x{}", item.getName(), item.getQuantity());
+                        } else {
+                            log.warn("\\tFailed to deposit individual equipment item: {} x{}", item.getName(), item.getQuantity());
+                        }
```
It should not be an issue here because we itemNotInItemReqsEquipment are retrieved Rs2Equipment, which should not return an item model with the wrong slot

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/VoxPlugins/schedulable/example/SchedulableExamplePlugin.java:
```diff
@@ -132,22 +232,74 @@ protected void startUp() {
+                config.lockConditionHotkey(),
+                config.testPreScheduleTasksHotkey(),
+                config.testPostScheduleTasksHotkey(),
+                config.cancelTasksHotkey());
```
we just log the value here, so we want to see if it is null -> not set

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/TaF/GiantSeaweedFarmer/GiantSeaweedFarmerPlugin.java:
```diff
@@ -1,127 +1,120 @@
+            }
+            Microbot.log("Stopping Giant Seaweed Farmer script due to soft stop event.");
+            Microbot.stopPlugin(this);
+        }
```
Your suggestion is not correct, we should ignore the event and not lockCondition.unlock(); we can not handle the post schedule task which here is simply stopping the plugin

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/tasks/requirements/requirement/SpellbookRequirement.java:
```diff
@@ -0,0 +1,385 @@
+            Microbot.log("Error fulfilling spellbook requirement " + getName() + ": " + e.getMessage());
+            return !isMandatory(); // Don't fail mandatory requirements due to exceptions
+        }
+    }
```
> for using the cancellation here we would need to implent a whatdog for the movemen similar  to the LocationRequirement and we dont need to check the context here because always ant to fullfill the spell book req.

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/pluginscheduler/SchedulerPlugin.java:
```diff
@@ -232,26 +243,66 @@ private void checkInitialization() {
+                    
+                    // Only disable non-essential, non-Microbot, non-schedulable plugins
+                    if (descriptor != null && !descriptor.enabledByDefault() && !(plugin instanceof SchedulablePlugin)) {
+                        try {
```
the // Only disable non-essential, non-Microbot, non-schedulable plugins was wrong, we want stop microbot plugin only.

Voxsylvae commented on docs/scheduler/api/schedulable-plugin.md:
````diff
@@ -241,30 +241,30 @@ When implementing the stop event handler:
 ```java
 // Inside PluginScheduleEntry.softStop()
 Microbot.getClientThread().runOnSeperateThread(() -> {
-    Microbot.getEventBus().post(new PluginScheduleEntrySoftStopEvent(plugin, ZonedDateTime.now()));
+    Microbot.getEventBus().post(new PluginScheduleEntryPostScheduleTaskEvent(plugin, ZonedDateTime.now()));
     return false;                
 });
 
````
runOnSeperateThread is not deprecated

Voxsylvae commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/agility/tasks/MicroAgilityPrePostScheduleTasks.java:
```diff
@@ -0,0 +1,173 @@
+                        log.warn("Magic level {} is below 55 required for High Alchemy. Alchemy disabled.", magicLevel);
+                        Microbot.getConfigManager().setConfiguration(agilityPluginConfigDescripor.getGroup().toString(), "alchemy", (boolean)false);
+                        // Note: We could disable alchemy in config here, but that might not be desired
+                    }
```
this.plugin.getConfigDescriptor() should not return null, because we override it

Voxsylvae committed:
feat: forestry event completion tracking and scheduler enhancements

Features:
- Add thread-safe forestry event completion counter system
- Implement completion tracking across all 9 forestry events
- Display completion count in woodcutting overlay when enabled
- Extract plugin management utilities to SchedulerPluginUtil
- Enhance login script with member world validation

Bug fixes:
- Fix concurrent cache update strategy execution issues
- Resolve TaskContext enum reference from IMMEDIATE to BOTH
- Correct requirement priority values in examples
- Fix login authentication error handling and state transitions
- Improve break handler integration and plugin lifecycle management

Voxsylvae commented on docs/scheduler/tasks/requirements-integration.md:
```diff
@@ -0,0 +1,259 @@
+
+Study these production examples for integration insights:
+
+**GOTR Requirements**: Examine the [GOTR requirements implementation](../../../runelite-client/src/main/java/net/runelite/client/plugins/microbot/runecrafting/gotr/) for sophisticated minigame requirement coordination.
+
+**Example Integration**: Review the [complete example](../../../runelite-client/src/main/java/net/runelite/client/plugins/microbot/VoxPlugins/schedulable/example/) for comprehensive requirement integration patterns.
+
+### Integration Patterns
```
The example links are pointing to the correct implementation, which I want to point to... we want to reference the implementation in the folder, which is an example implementation of how we can make a plugin schedulable—a more enhanced version of it with multiple examples via configurable conditions and pre/post scheduled tasks

Voxsylvae commented on docs/scheduler/tasks/README.md:
```diff
@@ -0,0 +1,123 @@
+
+Study these production codebases for practical implementation insights:
+
+- **GOTR Integration**: Examine [GOTR Plugin Implementation](../../../runelite-client/src/main/java/net/runelite/client/plugins/microbot/runecrafting/gotr/) demonstrating sophisticated minigame automation with comprehensive preparation requirements
+- **Example Plugin**: Review [SchedulableExample Implementation](../../../runelite-client/src/main/java/net/runelite/client/plugins/microbot/VoxPlugins/schedulable/example/) showcasing complete system integration patterns
+
```
The example links are pointing to the correct implementation, which I want to point to... we want to reference the implementation in the folder, which is an example implementation of how we can make a plugin schedulable—a more enhanced version of it with multiple examples via configurable conditions and pre/post scheduled tasks

Voxsylvae committed:
feat: comprehensive inventory management and forestry system improvements

This commit introduces major enhancements to inventory management, ground item handling, and forestry event system with performance optimizations and bug fixes.

- **Advanced inventory dropping system**: Added Rs2Inventory.dropAmount() methods with quantity control and InteractOrder support (EFFICIENT_ROW, COLUMN, ZIGZAG, etc.)
- **Intelligent inventory space management**: AutoWoodcuttingPlugin.ensureInventorySpace() for forestry events with smart log dropping
- **Lazy loading for Rs2GroundItemModel**: Implemented ItemComposition lazy loading pattern similar to Rs2ItemModel to improve performance
- **Cache validation in Script base class**: Added Rs2CacheManager validation to prevent script execution before cache is loaded

- **Fixed HivesEvent bot-like behavior**: Complete rewrite with state management to prevent spam clicking and bot detection
  - Added beehive completion tracking with HashSet
  - Implemented proper waiting for beehive completion (15-30 second timeouts)
  - Added natural delays between interactions (1-3 seconds)
  - Support for cooperative multiplayer (handles beehives completed by others)
- **Fixed ritual circle NPC tracking**: Changed from NPC to Rs2NpcModel for proper caching and lifecycle management
- **Fixed forestry event inventory management**: All events now ensure proper inventory space before starting

- **Rs2GroundItemModel optimization**: Lazy loading of ItemComposition reduces memory usage and initialization time
- **Improved error handling**: All composition-dependent methods now have try-catch guards with meaningful fallbacks
- **Better resource management**: Forestry events stop walking/navigation when active to prevent conflicts

- **Forestry event inventory requirements**:
  - EggEvent: Maintains existing space management
  - EntlingsEvent: Ensures 2 slots for tree leaves and potential rewards
  - FlowersEvent: Ensures 3 slots for pollen and fruits/seeds
  - FoxEvent: Ensures 1 slot for potential fox whistle
  - RitualEvent: Ensures 1 slot for potential petal garland
  - StrugglingSaplingEvent: Ensures 5 slots for various rewards
- **Enhanced forestry event lifecycle**: All events now call Rs2Walker.setTarget(null) to stop conflicting navigation
- **Improved logging and debugging**: Better error messages and state tracking throughout the system

See1Duck commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/aiofighter/constants/Constants.java:
Why was this file re-added?

Voxsylvae committed:
(feat,bugfixes,core): cache architecture overhaul and comprehensive pre/post schedule tasks system

- Renamed ScheduleContext to TaskContext for better clarity across requirements system
- Updated all requirement classes to use ExecutionResult enum instead of boolean success states
- Refactored cache management with thread-safe AtomicReference patterns

- Rs2CacheManager: Thread-safe profile management with AtomicReference
- Rs2Bank: Enhanced cache validation with profile consistency checks
- Rs2BankData: Improved cache loading with retry mechanisms
- Rs2ItemModel: Enhanced noted/unnoted item ID resolution

- SchedulerPlugin: Automatic disabling of non-essential plugins during scheduler startup
- SchedulerState: Enhanced state management for plugin lifecycle
- PluginScheduleEntry: Improved execution tracking and failure handling

- NEW: Granular execution result states (SUCCESS, SOFT_FAILURE, HARD_FAILURE)
- NEW: Enhanced failure tracking with retry capability detection
- NEW: Standardized result reporting across all scheduler events

- AbstractPrePostScheduleTasks: Complete rewrite with cancellation support
- TaskExecutionState: Comprehensive state tracking for task execution phases
- PrePostScheduleTasksOverlayComponents: Real-time task progress visualization

- ItemRequirement: Auto-resolution of noted items for stackable requirements
- LocationRequirement: Enhanced location validation with radius-based checks
- ShopRequirement: World hopping support with excluded worlds tracking (⚠️ ALPHA)
- LootRequirement: Collection-based requirement validation (⚠️ ALPHA)
- OrRequirement: Enhanced OR logic with multiple requirement modes

- NEW: Automatic InventorySetup generation from requirements
- NEW: Integration with MInventorySetupsPlugin for seamless setup creation
- NEW: Equipment and inventory slot planning with priority-based selection

- NEW: Default implementations for pre/post schedule task event handlers
- NEW: ExecutionResult-based reporting methods
- NEW: Enhanced lock condition management for critical operations
- DEPRECATED: Boolean-based success reporting methods

- NEW: Comprehensive requirement validation summaries
- NEW: Context-aware requirement filtering (PRE_SCHEDULE vs POST_SCHEDULE)
- NEW: Enhanced requirement solver with conflict resolution

- MicroAgilityPlugin: Full integration with pre/post schedule tasks
- GotrPlugin: Enhanced requirement validation and task management
- SchedulableExamplePlugin: Updated to demonstrate new features

- Fixed cache corruption during profile switches
- Resolved race conditions in concurrent task execution
- Fixed inventory setup creation with proper equipment slot assignments
- Corrected requirement fulfillment validation edge cases

- Shop and Collection requirements are in ALPHA state - report bugs via Discord
- InventorySetupPlanner is in BETA state - expect refinements
- All breaking changes maintain backward compatibility through deprecated methods

Voxsylvae committed:
feat: enhanced woodcutting system with comprehensive fletching support and log basket management

- Implement comprehensive Rs2Fletching utility with arrow, bolt, dart crafting support
- Add Rs2LogBasket utility for efficient log inventory management with bank integration
- Refactor woodcutting system with primary/secondary action enums and improved forestry event handling
- Enhance scheduler plugin system with improved lock condition handling and initialization reliability
- Add null safety checks and improved error handling across loot requirement and player location validation
- Streamline agility and runecrafting plugin scheduler mode detection using centralized isScheduleMode method
- Fix break handler import cleanup and improve deposit operations with proper timing validation

chsami committed:
Merge branch 'development' into PR_SchedulerPlugin_PrePostScheduleTasks
